torch/optim/lr_scheduler.py

```diff
@@ -0,0 +1,181 @@
+
+
+class ExponentialLR(LambdaLR):
+    """Set the learning rate to the initial LR decayed by gamma in
+    every epoch."""
+
+    def __init__(self, optimizer, base_lr, gamma):
+        super(ExponentialLR, self).__init__(optimizer, base_lr,
+                                            lambda epoch: gamma ** epoch)
+
+
+class ReduceLROnPlateau(object):
+    """Reduce learning rate when a metric has stopped improving.
+    Models often benefit from reducing the learning rate by a factor
+    of 2-10 once learning stagnates. This scheduler reads a metrics
+    quantity and if no improvement is seen for a 'patience' number
+    of epochs, the learning rate is reduced.
+    
+    Args:
+        factor: factor by which the learning rate will
+            be reduced. new_lr = lr * factor
+        patience: number of epochs with no improvement
+            after which learning rate will be reduced.
+        verbose: int. 0: quiet, 1: update messages.
+        mode: one of {min, max}. In `min` mode,
+            lr will be reduced when the quantity
+            monitored has stopped decreasing; in `max`
+            mode it will be reduced when the quantity
+            monitored has stopped increasing.
+        epsilon: threshold for measuring the new optimum,
+            to only focus on significant changes.
+        cooldown: number of epochs to wait before resuming
+            normal operation after lr has been reduced.
+        min_lr: lower bound on the learning rate.
+        
+        
+    Example:
+        >>> optimizer = torch.optim.SGD(model.parameters(), lr=0.1, momentum=0.9)
+        >>> scheduler = ReduceLROnPlateau(optimizer, 'min')
+        >>> for epoch in range(10):
+        >>>     train(...)
+        >>>     val_loss = validate(...)
+        >>>     # different from LambdaLR, step should be called after validate()
+        >>>     scheduler.step(epoch, val_loss) 
+    """
+
+    def __init__(self, optimizer, mode='min', factor=0.1, patience=10,
+                 verbose=0, epsilon=1e-4, cooldown=0, min_lr=0):
+
+        if factor >= 1.0:
+            raise ValueError('ReduceLROnPlateau '
+                             'does not support a factor >= 1.0.')
+        self.factor = factor
+        self.min_lr = min_lr
+        self.epsilon = epsilon
+        self.patience = patience
+        self.verbose = verbose
+        self.cooldown = cooldown
+        self.cooldown_counter = 0  # Cooldown counter.
+        self.monitor_op = None
+        self.wait = 0
+        self.best = 0
+        self.mode = mode
+        assert isinstance(optimizer, Optimizer)
+        self.optimizer = optimizer
+        self._reset()
+
+    def _reset(self):
+        """Resets wait counter and cooldown counter.
+        """
+        if self.mode not in ['min', 'max']:
+            raise RuntimeError('Learning Rate Plateau Reducing mode %s is unknown!')
+        if self.mode == 'min':
+            self.monitor_op = lambda a, b: np.less(a, b - self.epsilon)
```

Comment:
ReduceLROnPlateau is ported from Keras. But yes, this sounds reasonable and I will make an option for it.